mithril-common/src/test_utils/certificate_chain_builder.rs

```diff
@@ -0,0 +1,752 @@
+        let genesis_protocol_message = CertificateGenesisProducer::create_genesis_protocol_message(
+            next_protocol_parameters,
+            &next_avk,
+            &epoch,
+        )
+        .unwrap();
+        let genesis_signature = genesis_producer
+            .sign_genesis_protocol_message(genesis_protocol_message)
+            .unwrap();
+
+        CertificateGenesisProducer::create_genesis_certificate(
+            certificate.metadata.protocol_parameters,
+            certificate.metadata.network,
+            certificate.epoch,
+            #[allow(deprecated)]
+            certificate.metadata.immutable_file_number,
+            next_avk,
+            genesis_signature,
+        )
+        .unwrap()
+    }
+
+    fn build_standard_certificate(&self, context: &CertificateChainBuilderContext) -> Certificate {
+        let fixture = context.fixture;
+        let mut certificate = self.build_base_certificate(context);
+        certificate.metadata.signers = fixture.stake_distribution_parties();
+        let mut protocol_message = certificate.protocol_message.clone();
+        protocol_message.set_message_part(
+            ProtocolMessagePartKey::SnapshotDigest,
+            format!("digest-{}", context.index_certificate),
+        );
+        certificate.protocol_message = protocol_message;
+        certificate.signed_message = certificate.protocol_message.compute_hash();
+        let single_signatures = fixture
+            .signers_fixture()
+            .iter()
+            .filter_map(|s| {
+                s.protocol_signer
+                    .sign(certificate.signed_message.as_bytes())
+            })
+            .collect::<Vec<_>>();
+        let clerk = CertificateChainBuilder::compute_clerk_for_signers(&fixture.signers_fixture());
+        let multi_signature = clerk
+            .aggregate(&single_signatures, certificate.signed_message.as_bytes())
+            .unwrap();
+        certificate.signature = CertificateSignature::MultiSignature(
+            certificate.signed_entity_type(),
+            multi_signature.into(),
+        );
+
+        certificate
+    }
+
+    // Returns the chained certificates in reverse order
+    // The latest certificate of the chain is the first in the vector
+    fn compute_chained_certificates(&self, certificates: Vec<Certificate>) -> Vec<Certificate> {
+        let mut certificates_chained: Vec<Certificate> = Vec::new();
+        certificates
+            .iter()
+            .enumerate()
+            .for_each(|(i, certificate)| {
+                let mut certificate_new = certificate.clone();
+                if i > 0 {
+                    if let Some(previous_certificate) = certificates_chained.get(i - 1) {
+                        certificate_new.previous_hash = previous_certificate.compute_hash();
+                    }
+                } else {
+                    certificate_new.previous_hash = "".to_string();
+                }
```

Comment:
Yes, it would be better indeed. It will be fixed in a slightly different way with the implementation of this proposition: https://github.com/input-output-hk/mithril/pull/1999#discussion_r1802998682 :+1: